File: Ejemplos/38-DesayunoAsincronoJava/src/main/java/dev/joseluisgs/sync/DesayunoSync.java
package dev.joseluisgs.sync;

import dev.joseluisgs.alimentos.*;

import java.util.concurrent.TimeUnit;

public class DesayunoSync {

    public static void main(String[] args) throws InterruptedException {

        var timeInit = System.currentTimeMillis();

        var taza = servirCafe();
        System.out.println("El café está listo");

        var huevos = freirHuevos(2);
        System.out.println("Los huevos están listos");

        var beicon = freirBeicon(3);
        System.out.println("El beicon está listo");

        var tostada = tostarPan(2);
        ponerMantequilla(tostada);
        ponerMermelada(tostada);
        System.out.println("Las tostadas están listas");

        var zumoNaranja = servirZumoNaranja();
        System.out.println("El zumo de naranja está listo");
        System.out.println("¡El desayuno está listo!");

        var timeEnd = System.currentTimeMillis();

        System.out.println("Tiempo total: " + (timeEnd - timeInit) + " ms");
    }

    private static Zumo servirZumoNaranja() {
        System.out.println("Sirviendo zumo de naranja");
        return new Zumo();
    }

    private static void ponerMermelada(Tostada tostada) {
        System.out.println("Poniendo mermelada en la tostada");
    }

    private static void ponerMantequilla(Tostada tostada) {
        System.out.println("Poniendo mantequilla en la tostada");
    }

    private static Tostada tostarPan(int rebanadas) {
        for (int rebanada = 0; rebanada < rebanadas; rebanada++) {
            System.out.println("Poniendo una rebanada de pan en la tostadora");
        }
        System.out.println("Comenzando a tostar...");
        try {
            TimeUnit.SECONDS.sleep(3); // Simula la espera
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println("sacando la tostada de la tostadora");

        return new Tostada();
    }

    private static Beicon freirBeicon(int lonchas) throws InterruptedException {
        System.out.println("Poniendo " + lonchas + " lonchas de beicon en la sartén");
        System.out.println("Cocinando la primera loncha del beicon...");
        try {
            TimeUnit.SECONDS.sleep(3); // Simula la        } catch (InterruptedException e) {
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        for (int rebanada = 0; rebanada < lonchas; rebanada++) {
            System.out.println("Dando la vuelta a una loncha de beicon");
        }
        System.out.println("Cocinando el segundo loncha del beicon...");
        try {
            TimeUnit.SECONDS.sleep(3); // Simula la espera
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println("Poniendo el baicon en el plato");

        return new Beicon();
    }

    private static Huevo freirHuevos(int cantidad) {
        System.out.println("Calentando la sartén para los huevos...");
        try {
            TimeUnit.SECONDS.sleep(3); // Simula la espera
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println("Rompiendo " + cantidad + " huevos");
        System.out.println("Cocinando los huevos...");
        try {
            TimeUnit.SECONDS.sleep(3); // Simula la espera
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println("Poniendo los huevos en el plato");

        return new Huevo();
    }

    private static Cafe servirCafe() {
        System.out.println("Sirviendo café");
        return new Cafe();
    }
}

File: Ejemplos/37-DesayunoAsincronoKotlin/src/main/kotlin/sync/DesayunoSync.kt
package dev.joseluisgs.sync

import dev.joseluisgs.alimentos.*
import java.util.concurrent.TimeUnit

// Estas clases están intencionalmente vacías para el propósito de este ejemplo. Son simplemente clases marcadoras
// para el propósito de la demostración, no contienen propiedades y no sirven para otro propósito.


fun main(args: Array<String>) {
    val timeInit = System.currentTimeMillis()

    val taza = servirCafe()
    println("El café está listo")

    val huevos = freirHuevos(2)
    println("Los huevos están listos")

    val beicon = freirBeicon(3)
    println("El beicon está listo")

    val tostada = tostarPan(2)
    ponerMantequilla(tostada)
    ponerMermelada(tostada)
    println("Las tostadas están listas")

    val zumoNaranja = servirZumoNaranja()
    println("El zumo de naranja está listo")
    println("¡El desayuno está listo!")

    val timeEnd = System.currentTimeMillis()
    println("Tiempo total: ${timeEnd - timeInit} ms")
}

private fun servirZumoNaranja(): Zumo {
    println("Sirviendo zumo de naranja")
    return Zumo()
}

private fun ponerMermelada(tostada: Tostada) {
    println("Poniendo mermelada en la tostada")
}

private fun ponerMantequilla(tostada: Tostada) {
    println("Poniendo mantequilla en la tostada")
}

private fun tostarPan(rebanadas: Int): Tostada {
    for (rebanada in 0..<rebanadas) {
        println("Poniendo una rebanada de pan en la tostadora")
    }
    println("Comenzando a tostar...")
    try {
        TimeUnit.SECONDS.sleep(3) // Simula la espera
    } catch (e: InterruptedException) {
        e.printStackTrace()
    }
    println("sacando la tostada de la tostadora")

    return Tostada()
}

private fun freirBeicon(lonchas: Int): Beicon {
    println("Poniendo $lonchas lonchas de beicon en la sartén")
    println("Cocinando la primera loncha del beicon...")
    try {
        TimeUnit.SECONDS.sleep(3) // Simula la espera
    } catch (e: InterruptedException) {
        e.printStackTrace()
    }
    for (rebanada in 0..<lonchas) {
        println("Dando la vuelta a una loncha de beicon")
    }
    println("Cocinando el segundo loncha del beicon...")
    try {
        TimeUnit.SECONDS.sleep(3) // Simula la espera
    } catch (e: InterruptedException) {
        e.printStackTrace()
    }
    println("Poniendo el beicon en el plato")

    return Beicon()
}

private fun freirHuevos(cantidad: Int): Huevo {
    println("Calentando la sartén para los huevos...")
    try {
        TimeUnit.SECONDS.sleep(3) // Simula la espera
    } catch (e: InterruptedException) {
        e.printStackTrace()
    }
    println("Rompiendo $cantidad huevos")
    println("Cocinando los huevos...")
    try {
        TimeUnit.SECONDS.sleep(3) // Simula la espera
    } catch (e: InterruptedException) {
        e.printStackTrace()
    }
    println("Poniendo los huevos en el plato")

    return Huevo()
}

private fun servirCafe(): Cafe {
    println("Sirviendo café")
    return Cafe()
}

